src_tc/despike/despike.man
NAME: Despike --  Performs an adaptive low-pass filter of the input image
 
SYNOPSIS: despike infile outfile 
 
DESCRIPTION: 
      This program uses kernel filtering to remove high frequency
    information from the input byte image.  Processing proceeds by placing a 
    3X3 kernel over each point in the image.  Then, the mean and deviation 
    of all values x in the kernel such that min <= x <= max are computed.
    If the pixel being processed is more than nstd deviations from the mean
    AND is more than tol from the mean, it is considered invalid.
    Invalid pixels are replaced based on the invld option given.
      If invld = 1, invalid pixels are replaced with 0.
      If invld = 2, invalid pixels are replaced with the kernel mean.

    Processing continues until all points in the image have been despiked.
    This version of despike automatically performs two passes through each
    image.  For the first pass the parameters are:
	min = 0, max = 255, nstd = 1.5, tol = 50.0, invld = 1
    For the second pass through the image the paramters are:
	min = 1, max = 255, nstd = 1.5, tol = 50.0, invld = 2
    After the second pass, the input file, infile.img, has been despiked
    and the results are written to disk as the output file, outfile.img.
 
PARAMETERS/OPTIONS:         
    min       minimum pixel value to include in kernel mean calculation
    max       maximum pixel value to include in kernel mean calculation
    invld     invalid pixel processing options:
                   1) replace with 0
                   2) replace with mean
    nstd      number of standard deviations a pixel value must
              be from the kernel mean to be considered invalid
    tol       minimum value a pixel must be from the kernel 
              mean in order to be considered invalid
 
ERROR MESSAGES:
    MESSAGE GIVEN:                              REASON:
    ----------------------------------------------------------------------
    Failure allocating memory                   Out of memory
    Failure opening file <file>                 Invalid File Name Given
    Failure reading from file <file>            Improper File Type Given
 
EXAMPLES:
         despike infile outfile
                 
             This example executes the despike algorithm on infile.img
         producing outfile.img.  The first pass uses invalid pixel option 1.
         Thus, invalid pixels will be replaced by the value 0.  The minimum
         and maximum values to operate on are 0 to 255, corresponding
         to a byte image.  The threshhold for the number of standard
         deviations that a pixel value can be from its kernel mean before
         being considered invalid is set to 1.5.  Also, the minimum a pixel
         value must be from the kernel mean to be considered invalid is set
         to 50.0.  The second pass uses invalid pixel option 2.  This means 
         invalid pixels will be replaced by the kernel mean of the point.
         This time the minimum and maximum values to consider in the 
         the processing are 1 to 255.  Again the number of standard 
         deviations allowable is 1.5, and a pixel is valid if it is less
         than 50.0 from the mean.
         
SEE ALSO:  terrcorr
 
HARWARE/SOFTWARE LIMITATIONS:

BUGS:	    None Known
 
AUTHOR:     Tom Logan, ASF
 
CONTACT:    Tom Logan, ASF
            E-mail: tlogan@iias.images.alaska.edu



src_tc/despike/despike.c
/****************************************************************
NAME: DESPIKE -- Performs an adaptive low-pass filter
 
SYNOPSIS:  despike infile outfile
 
DESCRIPTION:
      This algorithm uses kernel filtering to remove high frequency
    information from the input image.  Processing proceeds by placing a 
    3X3 kernel over each point in the image.  The mean and deviation of
    all values x in the kernel such that min <= x <= max are then computed.
    If the pixel being processed is more than nstd deviations from the mean
    AND is more than tol from the mean, it is considered invalid.
    Invalid pixels are replaced based on the invld option given.
      If invld = 1, invalid pixels are replaced with 0.
      If invld = 2, invalid pixels are replaced with the kernel mean.
    Processing continues until all points in the image have been despiked.
    The input file, infile.img, is thus despiked to produce the output file,
    outfile.img.
 
    Parameters:
       infile    input .img byte file
       outfile   output .img byte file
	 
EXTERNAL ASSOCIATES: terrcorr
 
FILE REFERENCES:
    NAME:                USAGE:
    ---------------------------------------------------------------
    infile.img           Input image to despike  
    outfile.img          Results of despiking  
 
PROGRAM HISTORY:
  Version       Date       Author             Request
  -------       ----       ------             -------
    1           11/84   J. LaVergne  
    2            1/87   K. Zanter     conversion to unix         
    3            3/89   K. Narayanan  conversion to c and 5.0 standards
    4            8/89   J. Marmon     Changes to work on PN9050
    5            7/94   T. Logan      Redesigned without LAS or TAE 
				      depencies for the Cray YMP
    6.0          7/94   T. Logan      Converted to Parallel for Cray-T3D
    6.1		 2/95	T. Logan      Scalable problem size
    7.0		 10/95  T. Logan      Port from T3D to Solaris
    7.1		 8/98   O. Lawlor     Despike non-square images.
 
HARDWARE/SOFTWARE LIMITATIONS:
   
ALGORITHM DESCRIPTION:

    Prepare column sums for start of file 
    For all lines in this section
      For all pixels in the current line
	if (min <= input pixel <= max) include input pixel in mean
	else  don't include input pixel in mean  
	if (((pixel-mean)^2 > nstd^2*variance) && (abs(pixel-mean) > tol))
	  if (invld == 1) output pixel = 0
	  if (invld == 2) output pixel = mean
	else output pixel = input pixel
      Reset column sums for next image line
 
ALGORITHM REFERENCES:
 
BUGS:
*******************************************************************************
* Copyright (c) 2004, Geophysical Institute, University of Alaska Fairbanks   *
* All rights reserved.                                                        *
*                                                                             *
* You should have received an ASF SOFTWARE License Agreement with this source *
* code. Please consult this agreement for license grant information.          *
*                                                                             *
*                                                                             *
*       For more information contact us at:                                   *
*                                                                             *
*	Alaska Satellite Facility	    	                              *
*	Geophysical Institute			www.asf.alaska.edu            *
*       University of Alaska Fairbanks		uso@asf.alaska.edu	      *
*	P.O. Box 757320							      *
*	Fairbanks, AK 99775-7320					      *
*									      *
******************************************************************************/
#include "asf.h"
#include <sys/times.h>
#include "ddr.h"

#define VERSION 7.1

int
main (argc,argv)
    int argc; char **argv;
{
double  nstd;                        /* number of standard deviations  */
double  tol;                         /* tolerance value                */
double  kersum;                      /* sum of data in the kernel      */
double  varsum;                      /* sum of squared data in kernel  */
double  mean;                        /* mean of kernel being processed */
double  variance;                    /* variance in kernel             */
double  mydiff;                      /* difference from kernel mean    */
double  *sum;                        /* array of kernel column sums    */ 
double  *sqsum;                      /* array of column square sums    */
short   *vsum;                       /* # valid pixels in kernel column*/
int    divv;			     /* kernel sum divisor	       */
int    i, j, k, l;	             /* temporary storage	       */
int    line;                        /* loop counter                   */
int    del_line, add_line;          /* line to del/add to column sums */
int    invld;                       /* invalid pixel option           */
int    klines, ksamps;              /* no. lines & samples in kernel  */
int    nl, ns;                      /* no. lines & samples in image   */
int    valsum;                      /* no. elements in the kernel     */
int    xhalf, yhalf;                /* half kernel in x & y dimension */
unsigned char low;                   /* lower limit for data           */
unsigned char high;                  /* upper limit for data           */
unsigned char *in_buf;               /* input image                    */ 
unsigned char *out_buf;              /* output image                   */
unsigned char *temp;		     /* temp pointer for buffer swap   */
unsigned char *del_buf;              /* next line to del from kernel   */
unsigned char *add_buf;              /* next line to add to kernel     */
char    infile[255];	             /* name of input file	       */
char    outfile[255];		     /* name of output file	       */
char    inddr[255];	             /* name of input file ddr	       */
char    outddr[255];		     /* name of output file ddr	       */
char    cmd[255];		     /* buffer for system command      */
FILE    *fpin, *fpout;	             /* input and output file pointers */
int     loop;			     /* counter to determine pass      */
struct  DDR      ddr;                /* ddr structure                  */

StartWatch();

/* Get user entered parameters & echo 
 -----------------------------------*/
if (argc != 3)
 {
    printf("\nUsage: %s infile outfile\n",argv[0]);
    printf("       inputs:  infile     .img (LAS 6.0 BYTE image)\n");
    printf("      outputs:  outfile    .img (LAS 6.0 BYTE image)\n");
    printf("\n");
    printf("       Version %.2f,  ASF STEP TOOLS\n\n",VERSION);
    exit(1);
 }
strcat(strcpy(infile,argv[1]),".img");
strcat(strcpy(outfile,argv[2]),".img");
strcat(strcpy(inddr,argv[1]),".meta");
strcat(strcpy(outddr,argv[2]),".meta");
i = c_getddr(argv[1], &ddr);
if (i != 0) { printf("Unable to get ddr for file %s\n",infile); exit(1); }
nl = ddr.nl;
ns = ddr.ns;
printf("Processing Parameters:\n");
printf("\t input file %s is size %i*%i\n",infile,nl,ns);
printf("\t output file : %s\n",outfile);
 
/* Read the input file into memory 
 ------------------------------- */
in_buf  = (unsigned char *) MALLOC (nl*ns);
out_buf = (unsigned char *) MALLOC (nl*ns);

fpin = fopenImage(infile,"rb");
FREAD(in_buf,ns*nl,1,fpin);
fclose(fpin);
 
/* get remaining parameters
 -------------------------*/
klines = 3;
ksamps = 3;
xhalf = (ksamps - 1) / 2;
yhalf = (klines - 1) / 2;
 
sum   = (double *) MALLOC (ns*sizeof(double));
sqsum = (double *) MALLOC (ns*sizeof(double));
vsum  = (short *)  MALLOC (ns*sizeof(short));

for (loop = 0; loop < 2; loop++)
 {
  /* Set parameters for this loop */
  for (i = 0; i < nl*ns; i++) out_buf[i] = 0;
  if (loop == 0)
    {
      low = 0; high = 255; invld = 1; nstd = 1.5; tol = 50.0;
      printf("\t low = %i high = %i\n",low,high);
      printf("\t invld = %i, nstd = %f, tol = %f\n",invld,nstd,tol);
    }
  else 
    {
      low = 1; invld = 2; printf("\t low = %i high = %i\n",low,high);
      printf("\t invld = %i, nstd = %f, tol = %f\n",invld,nstd,tol);
    }
 
  /* prepare the initial kernel for filtering
   -----------------------------------------*/
  for (j = 0; j < ns; j++)
    {
      sum[j] = 0.0;
      vsum[j] = 0;
      sqsum[j] = 0.0;
    }
  for (line = 0; line <= xhalf ; line++)
    for (k = 0, l = line*ns; k < ns ; k++, l++)
      if (in_buf[l] >=  low  &&  in_buf[l] <=  high)
        {
          sum[k] += (double) in_buf[l];
          vsum[k]++;
          sqsum[k] += ((double)in_buf[l]*(double)in_buf[l]);
        }
  kersum = 0.0;
  valsum = 0;
  varsum = 0.0;
  for (k = 0; k <= xhalf ; k++)
   {
     kersum += sum[k];
     valsum += vsum[k];
     varsum += sqsum[k];
   }

  /* apply the low pass filter to the image
   -------------------------------------- */
  del_line = 0 - xhalf;
  del_buf  = &in_buf[0];
  add_line = xhalf + 1;
  add_buf  = &in_buf[add_line*ns];
 
  for (line = 0; line < nl; line++)
     {
       for (k = 0, l = line*ns; k < ns; k++,l++)
        {
  	  divv = valsum - 1;
	  if ((in_buf[l] < low) || (in_buf[l] > high)) divv = valsum;
	  if (divv  ==  0) mean = 0;
	  else mean = (kersum - (double)in_buf[l]) / divv;
	  if (valsum != 0)
	   variance = (varsum/valsum)-((kersum*kersum)/(valsum*valsum));
	  else variance = 0.0;
	  mydiff = in_buf[l] - mean;
	  if (( mydiff*mydiff > nstd*nstd*variance) && (fabs(mydiff) > tol ))
	   {
	     if (invld == 1) out_buf[l] = 0;
	     if (invld == 2) out_buf[l] = mean + 0.5;
	   }
	  else out_buf[l] = in_buf[l];
 
         /* reset kernel sum and sum valid values
  	  ------------------------------------- */
         if (k < ns-(xhalf+1))
  	   {
	     kersum += sum[k+xhalf+1];
	     valsum += vsum[k+xhalf+1];
	     varsum += sqsum[k+xhalf+1];
	   }
         if (k >= xhalf) 
	   {
	     kersum -= sum[k-xhalf];
	     valsum -= vsum[k-xhalf];
	     varsum -= sqsum[k-xhalf];
	   }
        } /* End of this line */
 
       if (del_line >= 0)
        {
  	  for (k = 0; k < ns ; k++)
	    if ( del_buf[k] >= low && del_buf[k] <= high)
	      {
	        sum[k] -= (double) del_buf[k];
	        vsum[k]--;
	        sqsum[k] -= ((double)del_buf[k]*(double)del_buf[k]);
	      }
	  del_buf += ns;
        }
       del_line++;
       if (add_line < nl)
        {
  	  for (k = 0; k < ns ; k++)
	    if (add_buf[k] >= low && add_buf[k] <= high)
	      {
	        sum[k] += (double) add_buf[k];
	        vsum[k]++; 
	        sqsum[k] +=  ((double)add_buf[k]*(double)add_buf[k]);
	      }
	  add_line++;
	  add_buf += ns;
        }
       varsum = 0.0;
       kersum = 0;
       valsum = 0.0;
       for (k = 0; k <= xhalf ; k++)
        {
  	  kersum += sum[k];
	  valsum += vsum[k];
	  varsum += sqsum[k];
        }
     }   /* End of a pass through the image */

    /* Swap input and output buffers for the next pass
     ------------------------------------------------*/
    temp = in_buf;
    in_buf = out_buf;
    out_buf = temp;  
  }
 
/* open, write, and close output image
 ------------------------------------ */
fpout = fopenImage(outfile,"wb");
fwrite(in_buf,nl*ns,1,fpout);
fclose(fpout);

sprintf(cmd,"cp %s %s\n",inddr,outddr);
if (system(cmd) != 0)
  { printf("Failure copying metadata: Program Exited\n\n"); exit(1); }

printf("\nDespike finished, wrote raw image %s\n",outfile);
StopWatch();

free(in_buf);
free(out_buf);
free(sum); 
free(sqsum);
free(vsum);

exit(0);
}

